pkg/services/ngalert/state/cache.go

```diff
@@ -201,23 +140,68 @@ func calculateState(ctx context.Context, log log.Logger, alertRule *ngModels.Ale
 	if len(dupes) > 0 {
 		log.Debug("Evaluation result contains either reserved labels or labels declared in the rules. Those labels from the result will be ignored", "labels", dupes)
 	}
+	return lbs, annotations
+}
 
-	cacheID := lbs.Fingerprint()
+func (c *cache) create(ctx context.Context, log log.Logger, alertRule *ngModels.AlertRule, result eval.Result, extraLabels data.Labels, externalURL *url.URL) *State {
+	lbs, annotations := expandAnnotationsAndLabels(ctx, log, alertRule, result, extraLabels, externalURL)
 
+	cacheID := lbs.Fingerprint()
 	// For new states, we set StartsAt & EndsAt to EvaluatedAt as this is the
 	// expected value for a Normal state during state transition.
 	newState := State{
-		AlertRuleUID:       alertRule.UID,
-		OrgID:              alertRule.OrgID,
-		CacheID:            cacheID,
-		Labels:             lbs,
-		Annotations:        annotations,
-		EvaluationDuration: result.EvaluationDuration,
-		StartsAt:           result.EvaluatedAt,
-		EndsAt:             result.EvaluatedAt,
-		ResultFingerprint:  result.Instance.Fingerprint(), // remember original result fingerprint
+		OrgID:                alertRule.OrgID,
+		AlertRuleUID:         alertRule.UID,
+		CacheID:              cacheID,
+		State:                eval.Normal,
+		StateReason:          "",
+		ResultFingerprint:    result.Instance.Fingerprint(), // remember original result fingerprint
+		LatestResult:         nil,
+		Error:                nil,
+		Image:                nil,
+		Annotations:          annotations,
+		Labels:               lbs,
+		Values:               nil,
+		StartsAt:             result.EvaluatedAt,
+		EndsAt:               result.EvaluatedAt,
+		ResolvedAt:           nil,
+		LastSentAt:           nil,
+		LastEvaluationString: "",
+		LastEvaluationTime:   result.EvaluatedAt,
+		EvaluationDuration:   result.EvaluationDuration,
+	}
+
+	existingState := c.get(alertRule.OrgID, alertRule.UID, cacheID)
+	if existingState == nil {
+		return &newState
+	}
+	// if there is existing state, copy over the current values that may be needed to determine the final state.
+	// TODO remove some unnecessary assignments below because they are overridden in setNextState
+	newState.State = existingState.State
```

Comment:
why not using [copy](https://github.com/grafana/grafana/pull/95985/files#diff-c9ddbec726aa73675318733e7d86ba72cdccbd68a8ad0d4f55930a6993fe56caR80) here? this could be a bit annoying if we'd need to add more fields to state

Reply:
Mostly due to performance, we calculate labels to get CacheID as well as annotations because they can change between evaluations. The copy method creates deep copies of those fields, which does not make sense in this case because they are overridden. 
Another reason is to keep it clear what fields are carried over to the "new state". This will likely go away in future refactorings.

Reply:
Will any test fail if we add a new field to the model but forget to copy it here? That would be great for catching issues early. Copy has a similar issue, I think, but it's closed to the model, harder to miss :)

Reply:
yes, there are couple of tests I added in this PR.